clang/test/SemaSYCL/builtin-fields.cpp

```diff
@@ -0,0 +1,230 @@
+  static_assert(__builtin_num_bases(decltype(l1)) == 0, "expected no bases");
+
+  int i;
+  auto l2 = [i](){};
+  static_assert(__builtin_num_fields(decltype(l2)) == 1, "expected one field");
+  static_assert(is_same<decltype(__builtin_field_type(decltype(l2), 0)), int>::value, "expected an int");
+
+  auto l3 = [&](){};
+  static_assert(__builtin_num_fields(decltype(l3)) == 0, "expected no fields");
+
+  auto l4 = [&]() { (void)i; };
+  static_assert(__builtin_num_fields(decltype(l4)) == 1, "expected one field");
+  static_assert(is_same<decltype(__builtin_field_type(decltype(l4), 0)), int&>::value, "expected an int&");
+
+  auto l5 = [=]() { (void)i; };
+  static_assert(__builtin_num_fields(decltype(l5)) == 1, "expected one field");
+  static_assert(is_same<decltype(__builtin_field_type(decltype(l5), 0)), int>::value, "expected an int");
+
+  // Test that invoking a generic lambda works with the correct number of
+  // captures.
+  double j;
+  invoker([=](auto X) { return i + j + X; });
+
+  // Test it still works even if there's odd stuff happening like using an
+  // if constexpr. Note: the order of the implicit captures is important to
+  // the static_assert in invoker(). Captures are converted into fields in the
+  // order they are implicitly captured.
+  invoker([=](auto X) {
+    if constexpr (is_same<decltype(X), int>::value) {
+      return (double)(X + i);
+    }
+    return (double)(X + j);
+  });
+}
+
+struct Z {
+  int x, y;
+  void lambdas() {
+    int z;
+    auto l1 = [this](){(void)(x + y); };
+    static_assert(__builtin_num_fields(decltype(l1)) == 1, "expected one field");
+    static_assert(__builtin_num_bases(decltype(l1)) == 0, "expected no bases");
+    static_assert(is_same<decltype(__builtin_field_type(decltype(l1), 0)), Z*>::value, "expected a Z*");
+    auto l2 = [&](){(void)(x + y); };
+    static_assert(__builtin_num_fields(decltype(l2)) == 1, "expected one field");
+    static_assert(is_same<decltype(__builtin_field_type(decltype(l2), 0)), Z*>::value, "expected a Z*");
+    auto l3 = [=](){(void)(x + y + z); };
+    static_assert(__builtin_num_fields(decltype(l3)) == 2, "expected two fields");
+    static_assert(is_same<decltype(__builtin_field_type(decltype(l3), 0)), Z*>::value, "expected a Z*");
+    static_assert(is_same<decltype(__builtin_field_type(decltype(l3), 1)), int>::value, "expected an int");
+  }
+};
+
+void errors() {
+  __builtin_num_fields(struct S); // expected-error {{'__builtin_num_fields' requires a complete type}} \
+                                     expected-note {{forward declaration of 'S'}}
+  __builtin_num_bases(struct T); // expected-error {{'__builtin_num_bases' requires a complete type}} \
+                                    expected-note {{forward declaration of 'T'}}
+
+  __builtin_num_fields(easy); // expected-error {{unknown type name 'easy'}}
+  __builtin_num_fields(decltype(easy)); // expected-error {{'__builtin_num_fields' requires a structure or lambda type}}

```

Comment:
Oh, that means you can introspect lambda captures?
Interesting...
I do not remember this is something done in the current ISO C++ SG7 proposal.
By the way, is this otherwise similar to the ISO C++ SG7 proposal prototype implementation?

Reply:
Yes, lambdas are just fancy structures as far as the AST is concerned, and the captures are fields of the structure.

I did not look at the SG7 proposal, so this work has no intentional relationship to that work.